pkg/queue/breaker.go

```diff
@@ -130,8 +135,17 @@ type semaphore struct {
 }
 
 // acquire receives the token from the semaphore, potentially blocking.
-func (s *semaphore) acquire() {
-	<-s.queue
+func (s *semaphore) acquire(timeout time.Duration) bool {
+	var timeoutCh <-chan time.Time
+	if timeout != 0 {
```

Comment:
this does not collect the channel, so it's unbounded memory leak on every activated request.
I suggest NewTimer() and then defer close().